File: 2015/day6/deno/logic_test.ts
import { assertEquals } from "https://deno.land/std@0.208.0/assert/mod.ts";
import { Action, translateInstructions, runInstructions, runInstructions2 } from "./logic.ts";

Deno.test("translateInstructions", () => {
  assertEquals(
    {
      action: Action.TurnOn,
      fromPositionStart: "0",
      fromPositionEnd: "0",
      toPositionStart: "999",
      toPositionEnd: "999"
    },
    translateInstructions("turn on 0,0 through 999,999")
  );
  assertEquals(
    {
      action: Action.Toggle,
      fromPositionStart: "0",
      fromPositionEnd: "0",
      toPositionStart: "999",
      toPositionEnd: "0"
    },
    translateInstructions("toggle 0,0 through 999,0")
  );
  assertEquals(
    {
      action: Action.TurnOff,
      fromPositionStart: "499",
      fromPositionEnd: "499",
      toPositionStart: "500",
      toPositionEnd: "500"
    },
    translateInstructions("turn off 499,499 through 500,500")
  );
});

Deno.test("runInstructions", () => {
  assertEquals(
    1000000,
    runInstructions("turn on 0,0 through 999,999\n")
  );
  assertEquals(
    1000,
    runInstructions("toggle 0,0 through 999,0\n")
  );
  assertEquals(
    999996,
    runInstructions("turn on 0,0 through 999,999\nturn off 499,499 through 500,500\n")
  );
});

Deno.test("runInstructions2", () => {
  assertEquals(
    1000000,
    runInstructions2("turn on 0,0 through 999,999\n")
  );
  assertEquals(
    2000,
    runInstructions2("toggle 0,0 through 999,0\n")
  );
  assertEquals(
    999996,
    runInstructions2("turn on 0,0 through 999,999\nturn off 499,499 through 500,500\n")
  );
  assertEquals(
    1,
    runInstructions2("turn on 0,0 through 0,0\n")
  );
  assertEquals(
    2000000,
    runInstructions2("toggle 0,0 through 999,999\n")
  );
});

File: 2015/day6/ruby/logic_test.rb
require 'test/unit'
require './logic'

class LogicTest < Test::Unit::TestCase
  def test_translate_instructions
    assert_equal(
      {
        "action" => "turn on",
        "fromPositionStart" => "0",
        "fromPositionEnd" => "0",
        "toPositionStart" => "999",
        "toPositionEnd" => "999"
      },
      translate_instructions("turn on 0,0 through 999,999")
    )
    assert_equal(
      {
        "action" => "toggle",
        "fromPositionStart" => "0",
        "fromPositionEnd" => "0",
        "toPositionStart" => "999",
        "toPositionEnd" => "0"
      },
      translate_instructions("toggle 0,0 through 999,0")
    )
    assert_equal(
      {
        "action" => "turn off",
        "fromPositionStart" => "499",
        "fromPositionEnd" => "499",
        "toPositionStart" => "500",
        "toPositionEnd" => "500"
      },
      translate_instructions("turn off 499,499 through 500,500")
    )
  end

  def test_run_instructions
    assert_equal(
      1_000_000,
      run_instructions("turn on 0,0 through 999,999\n")
    )
    assert_equal(
      1_000,
      run_instructions("toggle 0,0 through 999,0\n")
    )
    assert_equal(
      999_996,
      run_instructions("turn on 0,0 through 999,999\nturn off 499,499 through 500,500\n")
    )
  end

  def test_run_instructions2
    assert_equal(
      1_000_000,
      run_instructions2("turn on 0,0 through 999,999\n")
    )
    assert_equal(
      2_000,
      run_instructions2("toggle 0,0 through 999,0\n")
    )
    assert_equal(
      999_996,
      run_instructions2("turn on 0,0 through 999,999\nturn off 499,499 through 500,500\n")
    )
    assert_equal(
      1,
      run_instructions2("turn on 0,0 through 0,0\n")
    )
    assert_equal(
      2_000_000,
      run_instructions2("toggle 0,0 through 999,999\n")
    )
  end
end
